bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableIO.java

```diff
@@ -0,0 +1,1039 @@
+ * </pre>
+ *
+ * <p>
+ * To read a {@link PCollection} from a table, with an optional
+ * {@link Scan}, use {@link CloudBigtableIO#read(CloudBigtableScanConfiguration)}:
+ * </p>
+ *
+ * <pre>
+ * {@code
+ * PipelineOptions options = PipelineOptionsFactory.fromArgs(args).create();
+ * Pipeline p = Pipeline.create(options);
+ * PCollection<Result> = p.apply(
+ *   Read.from(CloudBigtableIO.read(
+ *      new CloudBigtableScanConfiguration.Builder()
+ *          .withProjectId("project-id")
+ *          .withInstanceId("instance-id")
+ *          .withTableId("table-id")
+ *          .build())));
+ * }
+ * </pre>
+ *
+ * <p>
+ * To write a {@link PCollection} to a table, use
+ * {@link CloudBigtableIO#writeToTable(CloudBigtableTableConfiguration)}:
+ * </p>
+ *
+ * <pre>
+ * {@code
+ * PipelineOptions options =
+ *     PipelineOptionsFactory.fromArgs(args).create();
+ * Pipeline p = Pipeline.create(options);
+ * PCollection<Mutation> mutationCollection = ...;
+ * mutationCollection.apply(
+ *   CloudBigtableIO.writeToTable(
+ *      new CloudBigtableScanConfiguration.Builder()
+ *          .withProjectId("project-id")
+ *          .withInstanceId("instance-id")
+ *          .withTableId("table-id")
+ *          .build()));
+ * }
+ * </pre>
+ */
+
+public class CloudBigtableIO {
+
+  enum CoderType {
+    RESULT,
+    RESULT_ARRAY;
+  }
+
+  private static AtomicCoder<Result> RESULT_CODER = new HBaseResultCoder();
+  private static final FlatRowAdapter FLAT_ROW_ADAPTER = new FlatRowAdapter();
+
+  /**
+   * A {@link BoundedSource} for a Cloud Bigtable {@link Table}, which is potentially filtered by a
+   * {@link Scan}.
+   */
+  @SuppressWarnings("serial")
+  static abstract class AbstractSource extends BoundedSource<Result> {
+    protected static final Logger SOURCE_LOG = LoggerFactory.getLogger(AbstractSource.class);
+    protected static final long SIZED_BASED_MAX_SPLIT_COUNT = 4_000;
+    static final long COUNT_MAX_SPLIT_COUNT= 15_360;
+
+    /**
+     * Configuration for a Cloud Bigtable connection, a table, and an optional scan.
+     */
+    private final CloudBigtableScanConfiguration configuration;
+
+    private transient List<SampleRowKeysResponse> sampleRowKeys;
+
+    AbstractSource(CloudBigtableScanConfiguration configuration) {
+      this.configuration = configuration;
+    }
+
+    @Override
+    public Coder<Result> getDefaultOutputCoder() {
+      return RESULT_CODER;
+    }
+
+    // TODO: Move the splitting logic to bigtable-hbase, and separate concerns between beam needs
+    // and Cloud Bigtable logic.
+    protected List<SourceWithKeys> getSplits(long desiredBundleSizeBytes) throws Exception {
+      desiredBundleSizeBytes = Math.max(getEstimatedSizeBytes(null) / SIZED_BASED_MAX_SPLIT_COUNT,
+        desiredBundleSizeBytes);
+      CloudBigtableScanConfiguration conf = getConfiguration();
+      byte[] scanStartKey = conf.getZeroCopyStartRow();
+      byte[] scanEndKey = conf.getZeroCopyStopRow();
```

Comment:
Shouldn't this be fetching the minimum start & end keys from the request rather then the first?